controllers/networkfence_controller.go

```diff
@@ -39,17 +72,90 @@ type NetworkFenceReconciler struct {
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the NetworkFence object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.10.0/pkg/reconcile
 func (r *NetworkFenceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx)
+
+	// fetch NetworkFence object instance
+	nwFence := &csiaddonsv1alpha1.NetworkFence{}
+	err := r.Get(ctx, req.NamespacedName, nwFence)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			logger.Info("NetworkFence resource not found")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// validate NetworkFence object so as its parameters are neither empty nor nil.
+	err = validateNetworkFenceSpec(nwFence)
+	if err != nil {
+		logger.Error(err, "failed to validate NetworkFence spec")
+
+		nwFence.Status.Result = csiaddonsv1alpha1.FencingOperationResultFailed
+		nwFence.Status.Message = fmt.Sprintf("Failed to validate Networkfence parameters: %v", util.GetErrorMessage(err))
+		statusErr := r.Client.Status().Update(ctx, nwFence)
+		if statusErr != nil {
+			logger.Error(statusErr, "Failed to update networkfence status")
+
+			return ctrl.Result{}, statusErr
+		}
+
+		// invalid parameter, do not requeue
+		return ctrl.Result{}, nil
+	}
+
+	logger = logger.WithValues("DriverName", nwFence.Spec.Driver, "CIDRs", nwFence.Spec.Cidrs)
+
+	conns := r.Connpool.GetByNodeID(nwFence.Spec.Driver, "")
+
+	// Iterate through the connections and find the one that matches the driver name
+	// provided in the NetworkFence spec; so that corresponding network fence and
+	// unfence operations can be performed.
+	for _, v := range conns {
+		for _, cap := range v.Capabilities {
+			// validate if NETWORK_FENCE capability is supported by the driver.
+			if cap.GetNetworkFence() == nil {
+				continue
+			}
+
+			// check if the networkfence object is getting deleted, if yes, then send
+			// UnfenceCluster request to the driver.
+			// After the UnfenceCluster request is processed, the finalizer will be removed
+			// so that the networkfence object can be deleted gracefully.
+			if !nwFence.GetDeletionTimestamp().IsZero() {
+				if util.ContainsInSlice(nwFence.GetFinalizers(), networkFenceFinalizer) {
 
-	// TODO(user): your logic here
+					err := r.unfenceClusterNetwork(ctx, nwFence, v.Client, logger)
```

Comment:
there is no check for the client to support `identity.Capability_NetworkFence_NETWORK_FENCE`? maybe not all sidecars support that? You have a check at line 146 for this, should that check be moved up?